### ISSUE TITLE
[Feature Request] Built-in "async" test support

### ISSUE THREAD

**Issue by leycec** [pytest-dev/pytest#10111]
I'd like to politely request that `pytest` promote support for `async` tests to a first-class feature provided out-of-the-box by `pytest` itself.

Currently, defining `async` tests requires adding one or more downstream third-party `pytest` plugins (e.g., [`pytest-asyncio`](https://github.com/pytest-dev/pytest-asyncio), [`pytest-mock-resources`](https://pypi.org/project/pytest-mock-resources)) as mandatory test-time project dependencies. That's fine for heavyweight test suites that need the full laborious public APIs implemented by those dependencies.

But most test suites just want one-off `async` tests *without* leveraging those APIs:

```pytest
# This should just work.
async def test_async_sleep() -> None:
    '''
    Unit test exercising the standard :func:`asyncio.sleep` function.
    '''

    from asyncio import sleep
    await sleep(0)
```

## So What Do You Propose, Exactly?

Ideally, `pytest` itself should trivially support `async` tests out-of-the-box.

Surprisingly, this *is* trivial. I know this, because [the test suite for @beartype already trivially supports `async` tests *without* `pytest` plugins](https://github.com/beartype/beartype/blob/5c245a6d62a4881597caf9032ca4d07f21d48a95/beartype_test/conftest.py#L41). Ignoring docstring, comments, and blank lines, @beartype's hook function implementing `async` support is **exactly 19 lines of code.** Seriously.

Everyone, witness 19 lines of code to support `async` tests:

```python
from asyncio import get_event_loop_policy, new_event_loop, set_event_loop
from functools import wraps
from inspect import iscoroutinefunction
from pytest import hookimpl
@hookimpl(hookwrapper=True, tryfirst=True)
def pytest_pyfunc_call(pyfuncitem: 'Function') -> None:
    test_func = pyfuncitem.obj
    if iscoroutinefunction(test_func):
        @wraps(test_func)
        def test_func_synchronous(*args, **kwargs):
            event_loop_old = get_event_loop_policy().get_event_loop()
            event_loop_old.close()
            event_loop = new_event_loop()
            set_event_loop(event_loop)
            test_func_coroutine = test_func(*args, **kwargs)
            event_loop.run_until_complete(test_func_coroutine)
            event_loop.close()
        pyfuncitem.obj = test_func_synchronous
    yield
```

**That's it.** Of course... docstring, comments, and whitespace are kinda essential. 19 lines might be trivial, but we'd all like to remember how and why that actually works in ten years when Bob in accounting reluctantly takes over. Let's make Bob's life easier, too:

```python
# Same thing, but actually readable. This is why we code.
from asyncio import (
    get_event_loop_policy,
    new_event_loop,
    set_event_loop,
)
from functools import wraps
from inspect import iscoroutinefunction
from pytest import hookimpl

@hookimpl(hookwrapper=True, tryfirst=True)
def pytest_pyfunc_call(pyfuncitem: 'Function') -> None:
    '''
    Hook wrapper called immediately *before* calling the passed test function.

    Specifically, this hook wrapper:

    * If this test function is synchronous (i.e., declared with ``def``),
      preserves this test function as is.
    * If this test function is asynchronous (i.e., declared with ``async
      def``), wraps this test function in a synchronous wrapper function
      synchronously running this test function under an event loop uniquely
      isolated to this test function. For safety, each asynchronous test
      function is run under a new event loop.

    This wrapper wraps all non-wrapper ``pytest_pyfunc_call()`` hooks and is
    hopefully called *before* all wrapper ``pytest_pyfunc_call()`` hooks. See
    also `the official pytest hook wrapper documentation <hook wrapper_>`__.

    Parameters
    ----------
    pyfuncitem : Function
        :mod:`pytest` object encapsulating the test function to be run.

    .. _hook wrapper:
       https://docs.pytest.org/en/6.2.x/writing_plugins.html#hookwrapper-executing-around-other-hooks
    '''

    # Test function to be called by this hook.
    test_func = pyfuncitem.obj

    # If this test function is an asynchronous coroutine function (i.e.,
    # callable declared with "async def" containing *NO* "yield"
    # expressions)...
    if iscoroutinefunction(test_func):
        @wraps(test_func)
        def test_func_synchronous(*args, **kwargs):
            '''
            Closure synchronously calling the current asynchronous test
            coroutine function under a new event loop uniquely isolated to this
            coroutine.
            '''

            # Current event loop for the current threading context if any *OR*
            # create a new event loop otherwise. Note that the higher-level
            # asyncio.get_event_loop() getter is intentionally *NOT* called
            # here, as Python 3.10 broke backward compatibility by refactoring
            # that getter to be an alias for the wildly different
            # asyncio.get_running_loop() getter, which *MUST* be called only
            # from within either an asynchronous callable or running event
            # loop. In either case, asyncio.get_running_loop() and thus
            # asyncio.get_event_loop() is useless in this context. Instead, we
            # call the lower-level get_event_loop_policy().get_event_loop()
            # getter -- which asyncio.get_event_loop() used to wrap. *facepalm*
            #
            # This getter should ideally return "None" rather than creating a
            # new event loop without our permission if no loop has been set.
            # This getter instead does the latter, implying that this closure
            # will typically instantiate two event loops per asynchronous
            # coroutine test function:
            # * The first useless event loop implicitly created by this
            #   get_event_loop() call.
            # * The second useful event loop explicitly created by the
            #   subsequent new_event_loop() call.
            #
            # Since there exists *NO* other means of querying the current event
            # loop, we reluctantly bite the bullet and pay the drunken piper.
            event_loop_old = get_event_loop_policy().get_event_loop()

            # Close this loop, regardless of whether the prior get_event_loop()
            # call just implicitly created this loop, because the "asyncio" API
            # offers *NO* means of differentiating these two cases.
            event_loop_old.close()

            # New event loop isolated to this coroutine.
            #
            # Note that this event loop has yet to be set as the current event
            # loop for the current threading context. Explicit is better than
            # implicit.
            event_loop = new_event_loop()

            # Set this as the current event loop for this threading context.
            set_event_loop(event_loop)

            # Coroutine object produced by this asynchronous coroutine test
            # function. Technically, coroutine functions are *NOT* actually
            # coroutines; they're just syntactic sugar implemented as standard
            # synchronous functions dynamically creating and returning
            # asynchronous coroutine objects on each call.
            test_func_coroutine = test_func(*args, **kwargs)

            # Synchronously run a new asynchronous task implicitly scheduled to
            # run this coroutine, ignoring the value returned by this coroutine
            # (if any) while reraising any exception raised by this coroutine
            # up the call stack to pytest.
            event_loop.run_until_complete(test_func_coroutine)

            # Close this event loop.
            event_loop.close()

        # Replace this asynchronous coroutine test function with this
        # synchronous closure wrapping this test function.
        pyfuncitem.obj = test_func_synchronous

    # Perform this test by calling this test function.
    yield
```

That's never broken once in the year-and-a-half since I wrote it. If I could say the same about everything else I've coded, my head would still have hair. This is why I facepalm.

Meanwhile, `pytest-asyncio` has:

* [**27 open issues** and **181 closed issues**](https://github.com/pytest-dev/pytest-asyncio/issues).
* [An awkward **Apache-2.0 license** incompatible with GPLv2](https://github.com/pytest-dev/pytest-asyncio/blob/master/LICENSE). :face_exhaling: 

None of this is to detract from the phenomenal work `pytest-asyncio` does. `pytest-asyncio` *is* phenomenal. We should applaud that. I certainly do.

Nonetheless, if we can trivially do something that everyone wants in 19 lines from within `pytest` itself, then we should do that thing. Simplicity is the virtue that I propose we embrace and extend here.

## Please. Just. Stop.

Thanks *so* much to the virtuous @pytest-dev team! Your tireless commitment to Pythonic QA has made all of our lives and test suites sane. My gratitude is eternal.

**Comment by RonnyPfannschmidt** [comment]
while i understand the sentient, for good core integration we need to solve the following topics in a sustainable manner

1. async fixtures that work with the same event loops/nurseries than the test call (aka managing and owning the event loop (including in the xdist case)
2. async hooks (people will eventually need it, and it will color the codebase (pluggy support needed) potentially massive undertaking
3. keeping the integration working for much saner and more modern async tools like trio/curio

in particular correct integration with fixtures and thread-models can pose a challenges that are far more code intensive than a cut down variant that only runs async test functions for asyncio

**Comment by leycec** [comment]
**Absolutely.** Thanks so much for the rapid and thoughtful response! And might I say, what a stunning red hat that is you have there.

Let's see if we can't tackle these caltrops in the road of progress.

## Other Event Loop Paradigms: Definitely In-scope

Firstly, concern **3.** seems like the primary concern – and you're absolutely right.

The 19-line snippet above is mostly just to propel discussion. It's definitely *not* indicative of a final working solution. What's needed to preserve both forward- and backward-compatibility across both the existing `pytest` plugin and third-party event loop landscape is for standard `asyncio` support to be strictly opt-in via a simple `pytest` fixture or decorator.

Decorators are easier. For the sake of debate, here's a trivial refactoring that defines an equally concise `@pytest.with_asyncio` decorator:

```python
from asyncio import get_event_loop_policy, new_event_loop, set_event_loop
from collections.abc import Callable
from functools import wraps
from inspect import iscoroutinefunction
from pytest import hookimpl

def with_asyncio(test_func: Callable) -> None:
    if not iscoroutinefunction(test_func):
        return test_func

    @wraps(test_func)
    def test_func_synchronous(*args, **kwargs):
        event_loop_old = get_event_loop_policy().get_event_loop()
        event_loop_old.close()
        event_loop = new_event_loop()
        set_event_loop(event_loop)
        test_func_coroutine = test_func(*args, **kwargs)
        event_loop.run_until_complete(test_func_coroutine)
        event_loop.close()

    return test_func_synchronous
```

Completely untested, because lazy. Pretty sure that works as advertised, but uncertain. In any case, safely isolating `async` support to a test-specific `asyncio` event loop via a standard decorator pattern is clearly trivial. Given that, the above example `async` test would then be testable ala: 

```python
from pytest import with_asyncio

# This should just work.
@with_asyncio
async def test_async_sleep() -> None:
    '''
    Unit test exercising the standard :func:`asyncio.sleep` function.
    '''

    from asyncio import sleep
    await sleep(0)
```

## Async Fixtures and Hooks: Really Out-of-Scope

Next, concerns **1.** (`async` fixtures) and **2.** (`async` hooks) both seem out-of-band here. The perfect is the enemy of the good. Again, the goal is just to bootstrap baseline support for `async` tests. We can both admit that full-blown asynchronous fixture and hook support is beyond the scope of a 19-line code snippet.

Thankfully, neither are vital. `pytest-asyncio` already supports `async` fixtures (and possibly hooks as well) and has for many years; everyone who needs either is almost certainly already using `pytest-asyncio`. That's fine and exactly how things should be.

All `pytest` should do, for the moment, is provide baseline support for `async` tests. Anything beyond that is still within the domain of downstream plugins.

## Multithreading: Ain't Nobody Goin' There

Lastly, I'm *fairly* certain that even `pytest-asyncio` doesn't address any of your trailing concerns of "correct integration with fixtures and thread-models." Notably, [`pytest-asyncio` (and thus presumably the snippet above) more or less behaves as expected under a single-threading model but understandably falls down under a multithreading model that comingles `async` fixtures with thread boundary-crossing](https://github.com/pytest-dev/pytest-asyncio/pull/228).

`pytest-asyncio` is the prevailing state of the art. If `pytest-asyncio` refuses to talk about something, there's no particular reason we should either. Right? Right.

Again, that's fine. The goal is just to bootstrap baseline support for `async` tests. `async` fixtures is out-of-scope. In theory, the above decorator *should* implicitly support multithreading paradigms – but it doesn't especially matter either way.

Pure-Python multithreading outside of GUI frameworks (that nobody increasingly uses, because the GIL exists and desktop is basically dead) is exceedingly rare in 2022. It's multiprocessing and GPU compute all the way down the data science pipeline, really.

## tl;dr

The perfect is the enemy of the good. `pytest-asyncio` already exists. The goal is *not* to reproduce `pytest-asyncio` in `pytest` but simply to bootstrap baseline support for `async` tests. We can do this... because I already did this.

**Comment by RonnyPfannschmidt** [comment]
Then my point is more that practically pytest-asyncio is the sensible minimum baseline, because it manages event loops and fixtures, i strongly prefer recommending a plugin with a working baseline than including a baseline that's targeting a minimum that will practically be a major showstopper once the need of users expands to async fixtures and more

It's OK if hooks can't do for now, but async tests without async fixtures will likely breed bad practices and horrendous to debug errors due to Event loop differences and the need for hacks in fixtures

**Comment by Zac-HD** [comment]
There's also Trio and for a few people Curio, so conflating async with asyncio would do a disservice to many of our users.
